sound/soc/sof/sof-audio.c

```diff
@@ -11,6 +11,144 @@
 #include "sof-audio.h"
 #include "ops.h"
 
+static int sof_expand_route_list(struct snd_sof_pcm *spcm, int dir,
+				 struct snd_sof_widget *next_widget, int num_widgets)
+{
+	struct snd_sof_route_list **route_list = &spcm->stream[dir].route_list;
+	struct snd_sof_route_list *new_list;
+
+	new_list = krealloc(*route_list, struct_size(new_list, widgets, num_widgets), GFP_KERNEL);
+	if (!new_list)
+		return -ENOMEM;
+
+	new_list->num_widgets = num_widgets;
+	new_list->widgets[num_widgets - 1] = next_widget;
+	*route_list = new_list;
+	return 0;
+}
+
+static int sof_route_add_connected_widgets(struct snd_sof_dev *sdev,
+					   struct snd_sof_pcm *spcm,
+					   struct snd_sof_widget *host_widget, int dir)
+{
+	struct snd_sof_route_list **route_list = &spcm->stream[dir].route_list;
+	struct snd_sof_widget *next_widget = host_widget;
+	struct snd_sof_widget *swidget;
+	struct snd_sof_route *sroute;
+	const char *route_endpoint, *next_widget_name;
+	bool found_next_widget;
+	int ret;
+
+	/* allocate memory for route_list */
+	*route_list = kzalloc(struct_size(*route_list, widgets, 1), GFP_KERNEL);
+	if (!(*route_list))
+		return -ENOMEM;
+
+	/* init route_list */
+	(*route_list)->widgets[0] = host_widget;
+	(*route_list)->num_widgets = 1;
+
+	/* set pipeline widget */
+	list_for_each_entry(swidget, &sdev->widget_list, list)
+		if (swidget->id == snd_soc_dapm_scheduler) {
+			const struct sof_ipc_pipe_new *pipeline = swidget->private;
+
+			if (pipeline && pipeline->pipeline_id == host_widget->pipeline_id) {
+				(*route_list)->pipeline_widget = swidget;
+				break;
+			}
+		}
+	if (!(*route_list)->pipeline_widget) {
+		dev_err(sdev->dev, "error: cant find pipeline widget with pipeline ID: %d\n",
+			host_widget->pipeline_id);
+		return -ENOENT;
+	}
+
+	/* add connected widgets in the current direction */
+	while (next_widget) {
+		found_next_widget = false;
+		list_for_each_entry(sroute, &sdev->route_list, list) {
+			/* get endpoint and next widget name based on direction */
+			if (dir == SNDRV_PCM_STREAM_PLAYBACK) {
+				route_endpoint = sroute->route->source;
+				next_widget_name = sroute->route->sink;
+			} else {
+				route_endpoint = sroute->route->sink;
+				next_widget_name = sroute->route->source;
+			}
+
+			/* add next widget to the route_list if available */
+			if (!strcmp(route_endpoint, next_widget->widget->name)) {
+				struct snd_sof_widget *swidget = NULL;
+
+				(*route_list)->num_widgets++;
+				found_next_widget = true;
+
+				/* find widget by name */
+				list_for_each_entry(swidget, &sdev->widget_list, list) {
+					if (!strcmp(swidget->widget->name, next_widget_name))
+						break;
+				}
+				if (!swidget)
+					return -ENOENT;
+
+				next_widget = swidget;
+
+				ret = sof_expand_route_list(spcm, dir, next_widget,
+							    (*route_list)->num_widgets);
+				if (ret < 0)
+					return ret;
```

Comment:
shouldn't you break out of the route-scanning loop here? Can there be multiple matches?

Reply:
@lyakh yes there can be  multiple matches, consider a mixer or a mux, they can have multiple sources/sinks.

Reply:
yes, I thought about those, but still wasn't sure. A simple case like a mixer, mixing PCM A and B to DAI 0 will not create multiple matches. Say, you enter this function to build a pipeline for the PCM A. The above `strcmp()` will look for a route with the source name of PCM A (or the respective buffer to be precise). So, that will only be one match. Same with a mux. I think the only possibilities would be topologies with forks for a single PCM. E.g. a demux in a playback pipeline (like sof-apl-demux-pcm512x.m4) or a mux / mixer in a capture pipeline (if such topologies even exist). Have you tested your PR with such topologies?

Reply:
yes, I need to. right now im failing even with simple topologies. I think I have to re-think the strategy of how we save the route info. Stay tuned. working on changes right now